### DISCUSSION TITLE
nidr related issue installing 6.20.0 on Ubuntu 24.04

### DISCUSSION THREAD

**Discussion by ET-w** [snl-dakota/dakota#149] [Q&A] [answered]
Hi all, 

I am trying to install Dakota on the new Ubuntu.  I've downloaded the source file dakota-6.20.0.tar.gz and *think* Ive done things right according to the instructions in "Source Quick Start"

Things go wrong with the cmake step:
erik@erik-Yoga-7-16IRL8:~/dakota-build$ cmake -C $DAK_SRC/cmake/BuildDakotaCustom.cmake $DAK_SRC
loading initial cache file /home/erik/dakota-6.20.0//cmake/BuildDakotaCustom.cmake
-- Configuring Dakota with CMake 3.28.3
CMake Error at CMakeLists.txt:33 (message):
  Required source sub-directory
  /home/erik/dakota-6.20.0/packages/external/nidr does not exist; if using a
  Git repository, are the submodules checked out?

-- Configuring incomplete, errors occurred!

Any idea where I am going wrong? 

Thanks.

**Comment by jadamstephens** [comment, marked as answer]
It's likely you downloaded the wrong source package. GitHub unhelpfully produces a source archive named "Source code" from the contents of the repo, which does not contain all the necessary submodules. Instead, you should download the one with "src" in the name. [Here's a direct link.](https://github.com/snl-dakota/dakota/releases/download/v6.20.0/dakota-6.20.0-public-src-cli.tar.gz)

**Reply by thenectarine** [reply]
I can confirm this.

I saw the same thing 60 days ago on Ubuntu 20.04.

This did not work: dakota-6.20.0.tar.gz

This worked: dakota-6.20.0-public-src-cli.tar.gz

**Reply by ET-w** [reply]
Were there any other tricks?  This is definitely the right source code, but I've tried different things and keep getting errors related to Teuchos like:

/home/erik/dakota-6.20.0-public-src-cli/src/util/util_data_types.hpp:16:10: fatal error: Teuchos_ParameterList.hpp: No such file or directory
   16 | #include "Teuchos_ParameterList.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.

**Reply by thenectarine** [reply]
I have some old installation notes to compile Dakota on Ubuntu 20.04. The notes might be off since the notes were hastily made afterwards.

Side comment before continuing on. 

> If you are on Windows or Red Hat Linux 7 or 8, you will not need to compile. Pre-compiled versions are already available. Just download, extract and run Dakota. Note each Dakota version has a different set of pre-compiled versions.
> 
> Go here: https://github.com/snl-dakota/dakota/releases
> For a specific release, click Assets to list the full list of avaialble downloads.
> Examples that include the pre-compiled Dakota executable.
> dakota-6.16.0-public-rhel7.Linux.x86_64-gui_cli.tar.gz
> dakota-6.16.0-public-windows.Windows.x64-gui_cli.zip

**Dakota v6.20.0 Compiling Instructions for Ubuntu 20.04**

1. Do NOT use Cmake 3.16, which is the default on Ubuntu 20.04.
   My notes are a little vague on this, but I think cmake 3.17 and newer worked for me.
   
   Uninstall old cmake (I think it is the line below, needs confirmation)
   
```
sudo apt purge --autoremove cmake
```

Install this cmake
   
```
sudo apt install cmake-mozilla
```


   Confirm the new version

```
cmake --version
cmake version 3.27.9
CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

2. Download and extract `dakota-6.20.0-public-src-cli.tar.gz` ([link](https://github.com/snl-dakota/dakota/releases/download/v6.20.0/dakota-6.20.0-public-src-cli.tar.gz))

3. Make sure your OS has the necessary libraries for compiling.

```
sudo apt-get install gcc g++ gfortran cmake libboost-all-dev libblas-dev liblapack-dev libopenmpi-dev openmpi-bin gsl-bin libgsl-dev  perl libhdf5-dev
sudo apt install chrpath
```

If you do not have Python, install Python.

```
sudo apt-get install python
```

4. Compile

```
mkdir build
cd ./build
cmake ../dakota-6.20.0-public-src-cli/
make -j 4
sudo make install .
```
5. Compiling takes a while. On my system with Intel Xeon, it took about 15-25 mins. 

6. Test the Dakota executable. If you see a version number, that is a good sign. 

```
dakota --version

Dakota version 6.20 released May 15, 2024.
Repository revision ff549cf04 (2024-05-13) built Aug 18 2024 20:22:34.
```
7. Do a test run. If the test run works, you are good to go.

```
dakota ./test/dakota_uq_cantilever.in
```
Output
```
Dakota version 6.20 released May 15, 2024.
Repository revision ff549cf04 (2024-05-13) built Aug 18 2024 20:22:34.
Running serial Dakota executable in serial mode.
Start time: Fri Oct 25 09:58:15 2024

--------------------------------
Begin DAKOTA input file
./test/dakota_uq_textbook_lhs.in
--------------------------------
#@ p*: Label=FastTest
#@ p0: MPIProcs=2

# DAKOTA INPUT FILE - dakota_textbook_lhs.in
 
environment,
	  graphics

method,
        sampling,
	  samples = 100 seed = 1

[...]
   7.0000000000e-01   9.5466580276e-01  -1.6918824205e+00  -1.6918824205e+00
   8.0000000000e-01   9.9042521425e-01  -2.3426064283e+00  -2.3426064283e+00
   9.0000000000e-01   9.9862024600e-01  -2.9933304362e+00  -2.9933304362e+00
   1.0000000000e+00   9.9986581169e-01  -3.6440544441e+00  -3.6440544441e+00
Cumulative Distribution Function (CDF) for response_fn_3:
     Response Level  Probability Level  Reliability Index  General Rel Index
     --------------  -----------------  -----------------  -----------------
   0.0000000000e+00   1.4049510996e-05   4.1883497710e+00   4.1883497710e+00
   1.0000000000e-01   1.4100694878e-04   3.6312846751e+00   3.6312846751e+00
   2.0000000000e-01   1.0552701517e-03   3.0742195793e+00   3.0742195793e+00
   3.0000000000e-01   5.9153471776e-03   2.5171544834e+00   2.5171544834e+00
   4.0000000000e-01   2.4992671711e-02   1.9600893876e+00   1.9600893876e+00
   5.0000000000e-01   8.0304797654e-02   1.4030242917e+00   1.4030242917e+00
   6.0000000000e-01   1.9878775380e-01   8.4595919589e-01   8.4595919589e-01
   7.0000000000e-01   3.8633120941e-01   2.8889410004e-01   2.8889410004e-01
   8.0000000000e-01   6.0571615023e-01  -2.6817099581e-01  -2.6817099581e-01
   9.0000000000e-01   7.9538121665e-01  -8.2523609166e-01  -8.2523609166e-01
   1.0000000000e+00   9.1656037924e-01  -1.3823011875e+00  -1.3823011875e+00
-----------------------------------------------------------------

<<<<< Iterator local_reliability completed.
<<<<< Environment execution completed.
DAKOTA execution time in seconds:
  Total CPU        =   0.456336 [parent =   0.456238, child =    9.8e-05]
  Total wall clock =  0.0749779

```

**Reply by ET-w** [reply]
Thanks.  I still cant shake the Teuchos related error.

**Reply by thenectarine** [reply]
I looked at this a little more.

My conclusion is that Ubuntu 24.04 is too new and changed too many things, which rendered some applications incompatible out of the box. This is normal and happens to every OS out there, e.g. Windows, Red Hat Linux, etc. 

I believe anything is possible as long as there is time, so I started experimenting to see if I can get compiling to work. Unfortunately the time my experimentation took reached the "abort threshold." It is much faster to downgrade to Ubuntu 20.04 and compile there than to get compiling to work on Ubuntu 24.04. I think this is a fair compromise and preferred instead of spending hours or days trying to fix the issues.

Perhaps a wizard here can figure it out?

Ultimately, this is expected. On average, I see challenges like this get resolved in 1-2 years, probably because there are many things to modify and requires multiple individuals to address.

Another thing. I amended my previous installation instructions for Ubuntu 20.04 so the instructions are more reliable.

**Comment by ET-w** [comment]
That's kinda what I was thinking. Maybe I'll try RHEL 8
Thanks.

**Reply by thenectarine** [reply]
FYI.

I saw the error about Trilinos on Ubuntu 20.04. The solution was to use `sudo make install .` . I wonder if the missing sudo is why it did not compile on Ubuntu 24.04.

NOT OK
`make install .
`

OK
`sudo make install .
`

Error
```

-- Up-to-date: /usr/local/include/Trilinos_version.h
CMake Error at packages/external/trilinos/cmake_install.cmake:46 (file):
file INSTALL cannot set permissions on
"/usr/local/include/Trilinos_version.h": Operation not permitted.
Call Stack (most recent call first):
cmake_install.cmake:47 (include)


make: *** [Makefile:130: install] Error 1
```
